File: src/912.py
class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        #我理解了快排的思路了
        #现在我要动手重拳出击了
        #44%, 52%
        endIndex = len(nums)
        if endIndex <= 1:
            return nums
        else:
            #调用快排
            self.quickSort(nums, 0, endIndex - 1)
            return nums

    def quickSort(self, _nums, _startIndex, _endIndex):
        print(_nums)
        if _startIndex == _endIndex:
            #终止递归条件
            return
        else:
            #切分，返回切分位置并排序切分位置左右
            partitionIndex = self.partitionAndSort(_nums, _startIndex, _endIndex)
            #找到一个元素的位置后，向左右递归
            if partitionIndex > _startIndex:
                self.quickSort(_nums, _startIndex, partitionIndex - 1)
            elif partitionIndex < _endIndex:
                self.quickSort(_nums, partitionIndex + 1, _endIndex)

    def partitionAndSort(self, _nums, _startIndex, _endIndex):
        if len(_nums) == 0 or _startIndex < 0 or _endIndex >= len(_nums):
            #异常值
            return -1
        else:
            #首先，拿到切分点
            index = random.randint(_startIndex, _endIndex)
            #然后，把切分点的值放到最后去
            _nums[_endIndex], _nums[index] = _nums[index], _nums[_endIndex]
            #开始交换
            small = _startIndex - 1
            index = _startIndex
            #不要等于，等于是最后切分元素存放的位置了
            while index < _endIndex:
                #每个元素跟切分的元素相比较
                if _nums[index] < _nums[_endIndex]:
                    #小于，就进行操作
                    small += 1
                    if small != index:
                        #交换small和index
                        _nums[small], _nums[index] = _nums[index], _nums[small]
                index += 1
            small += 1
            #把切分元素放到自己的位置
            _nums[small], _nums[_endIndex] = _nums[_endIndex], _nums[small]
            return small

File: src/912.cpp
class Solution {
    //手写快排，成就达成
    //干翻美团
    //十五分钟，一次提交，时间击败-64.94%，空间击败-31.64%
private:
    int getPartition(vector<int> & _nums, int _left, int _right){
        //选定切分位置
        int pos = rand() % (_right - _left + 1);    //生成小于长度的随机数
        pos += _left;   //加上左边起始量
        swap(_nums[pos], _nums[_right]);    //库函数swap
        //对_left - (_right - 1)内排序
        return doPartition(_nums, _left, _right);
    }

    int doPartition(vector<int> & _nums, int _left, int _right){
        //开始排序
        int i = _left -  1;
        for(int j = _left; j < _right; ++ j){
            if(_nums[j] < _nums[_right]){
                //发现小的了，交换
                i ++;
                swap(_nums[j], _nums[i]);
            }
        }
        //结束以后，把主元放回自己的位置
        swap(_nums[i + 1],  _nums[_right]);
        return (i + 1);
    }

    void quickSort(vector<int> & _nums, int _left, int _right){
        //递归向下的函数
        if(_left <  _right){
            //选定切分位置，并且排好左右的序列
            int pos = getPartition(_nums, _left, _right);
            //递归向下
            quickSort(_nums, _left, pos - 1);
            quickSort(_nums, pos + 1, _right);
        }
        return;
    }
public:
    vector<int> sortArray(vector<int>& nums) {
        //兄弟们，给我冲排序数组
        //手写快排的时候到了
        //官方题解的快排还写得挺好理解的，开整
        srand(time(0));     //选定随机数种子
        quickSort(nums, 0, nums.size() - 1);    //开始快排
        return nums;    
    }
};
